### ISSUE TITLE
Think through more granular permission and user management

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#1093]

**Comment by bickelj** [comment]
To the extent possible, I think we should let Keycloak be canonical for users, groups, and permissions. In the worst case, that means syncing data from Keycloak on occasion into PDC with some data redundancy. In the best case, it means PDC exclusively uses Keycloak interfaces to manage users. Keycloak has no concept of our data model (nor should it), so this means we will need to reference Keycloak objects from PDC.

**Comment by bickelj** [comment]
Looking through [Keycloak authorization docs](https://www.keycloak.org/docs/25.0.1/authorization_services/), Keycloak seems to have the ability to know about our application resources and to protect them. This is the reverse of what I said above regarding Keycloak not having a concept of our data model. My sense at the moment is that if we have to make updates in Keycloak for each and every record within one of our entities (example: adding a new organization or proposal causes a new resource created/tracked in Keycloak) I would hesitate to take that route right away. We don't want to merely protect `/blah`, we want to assign ownership of `/blah/5` to one party and assign ownership of `/blah/6` to another, etc.

**Comment by slifty** [comment]
@bickelj in the last phase we began to adopt a philosophy of: "keycloak is responsible for roles, but PDC is responsible for permissions"

Which is to say, keycloak would have things like "superuser" in it, but PDC would be responsible for knowing "this account is { associated with} with { Insert PDC Domain Entity Here }"

For instance, in that model Keycloak wouldn't be the place where a given account gets associated as having ownership permissions of a given organization.

I *think* this is the correct approach?  You probably have a better sense than me.  It seems the less we force KeyCloak to undertand the PDC the better we are in the long run (e.g. we rely on KeyCloak for authentication, we rely on PDC for authorization)

**Comment by bickelj** [comment]
@slifty That appears to be the most sustainable and balanced approach, yes. However, I'm considering if it's feasible for "PDC is responsible for permissions" to mean "PDC uses a standard permissions API exposed by Keycloak to manage permissions for PDC." At the moment it is still looking too complex.

My exploration of such Keycloak maximization still leads me in the direction of "Users and roles (or groups or both) in Keycloak, interpretation of who can access what in PDC by looking at the JWT." I think that's what you're saying above.

In other words, your expectation is still "permission for resource X in the PDC service database is an attribute or relation in PDC service database" and NOT "permission for resource X in the PDC service database is an attribute or relation in the Keycloak database to be queried via API."

**Comment by bickelj** [comment]
There are a number of stories regarding "my" data and "my" organization already present. Examples:
https://github.com/orgs/PhilanthropyDataCommons/projects/4/views/6?sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=58424322&sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=Status&pane=issue&itemId=39313981
https://github.com/PhilanthropyDataCommons/meta/issues/14
https://github.com/PhilanthropyDataCommons/meta/issues/11

A significant one here: https://github.com/PhilanthropyDataCommons/meta/issues/48

**Comment by bickelj** [comment]
The two paths I still see are these:
1. Users, roles, and permissions all in Keycloak, managed by API calls from PDC.
2. Users and roles in Keycloak, ACLs or something like it on every permissionable row in PDC.

Both feel pretty ugly. I have a clearer idea of what the latter's initial implementation would look like in the PDC database, namely some `jsonb` column on every permissionable table in PDC, with some ACL-ish structure denoting which users or roles can do what in that JSON object. I have a vaguer idea of what the former's implementation would look like for using Keycloak, but it seems like a bunch of API calls to create permissions objects for every row created.

I tried to find some standard JSON schema for ACLs and there are some examples for S3 or other systems out there but I didn't find a clear standard. I like the "viewers, editors, owners" concept from https://cloud.google.com/storage/docs/gsutil/commands/acl, though. That makes sense to me.

**Comment by kfogel** [comment]
@slifty Please see [my Zulip message](https://chat.opentechstrategies.com/#narrow/stream/75-PDC/topic/permissions/near/200144) just now.

**Comment by kfogel** [comment]
Thanks to @slifty for an excellent conversation about this just now.  Here is a plan, based on that conversation:

I'll start out with a description of what I think we're aiming for, from a user perspective, and then address Keycloak's role in that separately afterwards.  I'll use "Changemaker" throughout to refer to what the entity model currently calls an "Organizations" (because we know that term will become "Changemaker" once issue #1117 is taken care of).

In the PDC, we have a lot of cases where parties are making assertions about other parties' data.  To take the most common example: Data Providers and Funders are often providing proposal data that ultimately comes from Changemakers (i.e., is an assertion that "this Changemaker sent someone this Proposal Version at some point" or something like that).

So we don't want to just have a very traditional ACL-ish system in which users have read/write/admin roles on data that is "owned" by various organizations that are represented in the system by organizational entities.  We do *have* those organizational entities within the PDC, but their access relationships with data is quite non-traditional.

So here's our plan.  Let's start with some guiding principles:

1) Users are the entities that authenticate to PDC (which they do through Keycloak).

2) A User is affiliated with one or more organizational entities in PDC: Funders, Changemakers, Data Providers, the PDC itself, and in the future perhaps other types of organizational entities.   (I suspect >90% of users will be affiliated with just one organizational entity, and you'll see that entity right there in their email address, but there's no reason not to support multiple affiliations per User -- we can write loops, after all.

3) Every action taken in the PDC is attributed to a User.  E.g., We don't say that Funder F did something; we say that User U did that thing (though maybe User U is affiliated with Funder F at the time they did the thing -- of course, we're logging changes in affiliation, so at least in theory we could later go back and reconstruct what affiliations were true *at the time an action was taken*, although we are not building a dedicated interface for that because it's not a question that comes up often anyway).

Now, who can do what with what and who can't?  Well:

* Any Changemaker can see Proposals associated with themselves, and create new Proposal Versions for those Proposals.

* Changemakers *cannot* see Proposal data that they are not associated with.

  (Sidebar: Exactly how are Proposals associated to Changemakers?  Well, because whoever created the proposal made the assertion, i.e., used the API to say that "this Proposal is associated, in an ownership / authorship sense, with this Changemaker".  IOW, this business logic does not live in the PDC, because it can't -- only the provider of the data can make this assertion, and so it is their responsibility to do so.  In our own bulk uploader, Changemaker-to-Proposal relationships are based on Tax ID: if a proposal has a tax ID, then we make the association, and if the proposal doesn't have a Tax ID, then the bulk uploader doesn't make the association -- and note that a Proposal is not strictly required to have a relationship to any Changemakers at all.  Again, this is an ownership/authorship relationship, not a "Source" entity relationship, which means something different in the PDC.)

* Any Funder or Data Provider can create a new Proposal Version associated with any Changemaker.

* Funders can view and edit only Applications associated with themselves (again, "associated" in the ownership/authorship sense).

* Only PDC Admin creates/edits base fields right now.  This is a restriction we're imposing in order to maintain metadata cleanly right now.  If the inconvenience of having PDC Admin involved in new base field creation becomes too high, we can revisit this decision (but I suspect we'll want to keep it as-is, because creation of new Applications is fairly rare, and we want to do field-unification whenever possible to keep the number of distinct base fields to a minimum).

What about user management?  Well:

* Any User U can create new users affiliated with with any organization that U is affiliated with.  Maybe in the future we'll decide we need two classes of user for every organzation: admin users, and regular users, with only the former able to do user management.  But our usage scenarios right now don't point toward that level of bureaucracy, and since the PDC keeps audit trails for everything anyway, let's try the simpler way first.

And what is the UI by which users perform user management?

Answer: probably that UI lives in the PDC itself.  Keycloak does not (and should not) know about user<->organizational connections; Keycloak just authenticates users.  And in any case, Keycloak's UI would be too complex and too general; PDC users should get a dedicated UI for user management.

**Comment by slifty** [comment]
This is a great set of use cases and scope!

I think we can declare the "issue" closed and I'll make a series of new issues that flesh out an implementation plan based on the above.  Hooray!

**Comment by kfogel** [comment]
@slifty I'm re-opening this because I realized some important functionality we need to support:

Changemakers also need to be able to specify that certain parties *cannot* see their stuff.  I think they could do this in two ways:

1) Opt-out basis -- "Everyone *except* Funders A, B, and Q can see my information."
2) Opt-in basis -- "No one can see my stuff, except Funders A, B, and Q can see it."

I have a notion that the straightforward way to implement this internally is to have a "default permissions mode", either (1) or (2), for each organization relative to each other type of organization.  In other words, each Changemaker would have a DPM for Funders, a possibly different DPM for other Changemakers, etc.  If a given DPM is opt-out (1), then we need to consult a block-list to determine whether a given request is allowed; if a given DPM is opt-in (2), then we need to consult an allow-list to determine if a given request is allowed.

I'm not sure this is the best implementation plan -- happy to discuss / defer, of course.  I offer it partly because it also suggests the UI by which organizations -- principally Changemakers of course, and our initial UI implementation of this might start with just them, but in principle there's no long-term reason to limit this degree of control to Changemakers -- control permissions.  /CC @reefdog

**Comment by slifty** [comment]
Oh great point -- we were only thinking about write permissions!

I'm trying to think about the use case of opt-out basis.  What are the use cases there?  I have some followup questions but they may well be either answered or made irrelevant based on those use cases!

**Comment by kfogel** [comment]
@slifty Oh, the use case has been very much discussed by the management team -- and has been brought up by the Changemakers advising us, too.  Sometimes a Changemaker doesn't want a *particular* funder to see a particular proposal, for example because it's too similar to something that funder has already funded.

That makes it sound nefarious, but it's not.  The point of offering this level of permissions knob is so that Changemakers can know that they *really are in control*.

**Comment by slifty** [comment]
@kfogel thanks that makes sense; if it was something where it was more important to actually ensure the data isn't ever seen or found by a given agent that would be more problematic!

I do think it's important to note that use case presumes a specific funder use case which may not be the only reason a funder is looking at proposal data -- for instance, if the funder is trying to run an analysis of the proposal landscape for statistical or research purposes then that omission might not actually be desired by the Changemaker.

**Comment by kfogel** [comment]
@slifty Oh, the funder should be able to see the correct, more recent address.  We're just talking about "sensitive" data, like a proposal's description or budget or something, not fields that reflect already publicly-available data (a.k.a., Form 990 data).

**Comment by slifty** [comment]
@kfogel got it.

Right now we don't have any features around being able to specify something as `sensitive` or `not sensitive` and so I think we are going to need to have a more detailed exploration of this feature set.  We *do* have the distinction of `organization` vs `proposal` data however.  We don't have any mechanism to know what organization data is considered "publicly available".

Org vs Proposal might be good enough, although I want to be sure it's clear that when we talk about provenance of a given field value we need to realize that we might say that a value came from a proposal that is not visible (even thought he value is visible) and that is a somewhat confusing concept.

Also, I think we should clarify the risk factor here for the change maker / what they are expecting if they mark a proposal as being hidden to just one entity (as opposed to being removed entirely from the PDC): what happens if the funder does in-advertantly end up seeing that sensitive data?  For instance, if a group of funders are working together on a project that involves working with PDC data, and Funder A (with access) downloads a corpus and works with Funder B who then has access to that proposal.

All of this said, I think it would be good to have a followup discussion to flesh out this space more concretely before we start implementation in October.

We're good to go on the "write" permission sets I think, which reflect a meaningful enough scope of work.  We can make good progress there as we hone in on the "read" sets.

**Comment by slifty** [comment]
We talked through this a bit, a few conclusions:

1. Karl will raise the question of "public for all except Foundation Foo" with the PDC team (aka the "opt out option").  Our hypothesis is that this won't actually be a valuable feature, and might even be misleading / provide a false sense of privacy to the end user.
2. There is further discussion about "sensitive" vs "not sensitive" -- we may need a enum value (or even a flag) for "not reviewed" as well.
3. There is also the question of "private" vs "public" which may be a worthwhile additional distinction to have.

**Comment by kfogel** [comment]
For the record, I have asked Mansha and @jmergy about this, as the second item in this email:

```
From: Karl Fogel
To: Mansha Siraj, Jonathan Mergy
Cc: Dan Schultz
Subject: Two questions about labeling and permissions in the PDC.
Date: Tue, 01 Oct 2024 14:22:04 -0500
Message-ID: <871q0z8w83.fsf@red-bean.com>
```

**Comment by kfogel** [comment]
Answer from @jmergy:

> I agree with Option 2 path for the reasons mentioned...

I.e., we implement the "opt-in basis" as the only restrictive choice: "No one can see my stuff, except Funders A, B, and Q can see it."  Of course, we _also_ implement the option of "Everyone can see my stuff"; we just don't implement the "opt-out" exceptions to that.

**Comment by slifty** [comment]
Thank you for the clarifications @kfogel (and @jmergy via email!)

I'm going to mark this as closed now since I think we have achieved the issue at hand.  Of course this doesn't mean we're done thinking, though who knows maybe by closing the issue I will end our capacity for thought... oh dear I hope not.

**Comment by slifty** [comment]
Why do feet... stop at the bottom?

**Comment by kfogel** [comment]
> Why do feet... stop at the bottom?

They don't!  They just get very attenuated.
